//    Copyright 2019 Google Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        https://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

syntax = "proto3";

package google.fhir.r4.core;

import "proto/annotations.proto";
import "proto/r4/core/codes.proto";
import "proto/r4/core/datatypes.proto";
import "proto/r4/core/extensions.proto";

option java_multiple_files = true;
option java_package = "com.google.fhir.r4.core";
option go_package = "github.com/google/fhir/go/proto/r4/core";
option (.google.fhir.proto.fhir_version) = R4;

// Auto-generated from StructureDefinition for DataElement constraint on
// ElementDefinition data type. Definition of an element in a resource or
// extension. See http://hl7.org/fhir/StructureDefinition/elementdefinition-de
message DataElementConstraintOnElementDefinitionDataType {
  option (.google.fhir.proto.structure_definition_kind) = KIND_COMPLEX_TYPE;
  option (.google.fhir.proto.fhir_profile_base) =
      "http://hl7.org/fhir/StructureDefinition/ElementDefinition";
  option (.google.fhir.proto.fhir_structure_definition_url) =
      "http://hl7.org/fhir/StructureDefinition/elementdefinition-de";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "min.empty() or max.empty() or (max = '*') or iif(max != '*', min <= max.toInteger())";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "contentReference.empty() or (type.empty() and defaultValue.empty() and fixed.empty() and pattern.empty() and example.empty() and minValue.empty() and maxValue.empty() and maxLength.empty() and binding.empty())";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "pattern.empty() or (type.count() <= 1)";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "fixed.empty() or (type.count()  <= 1)";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "binding.empty() or type.code.empty() or type.select((code = 'code') or (code = 'Coding') or (code='CodeableConcept') or (code = 'Quantity') or (code = 'string') or (code = 'uri')).exists()";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "sliceIsConstraining.exists() implies sliceName.exists()";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "pattern.empty() or fixed.empty()";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "constraint.select(key).isDistinct()";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "type.select(code).isDistinct()";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "sliceName.empty() or sliceName.matches('^[a-zA-Z0-9\\\\/\\\\-_\\\\[\\\\]\\\\@]+$')";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "defaultValue.empty() or meaningWhenMissing.empty()";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "isModifier implies isModifierReason.exists()";
  option (.google.fhir.proto.fhir_path_message_constraint) =
      "path.matches('[A-Za-z][A-Za-z0-9]*(\\\\.[a-z][A-Za-z0-9]*(\\\\[x])?)*')";

  // Unique id for inter-element referencing
  String id = 1;

  // Additional content defined by implementations
  repeated Extension extension = 2;

  // Extensions that cannot be ignored even if unrecognized
  repeated Extension modifier_extension = 3;

  // Path of the element in the hierarchy of elements
  String path = 4
      [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

  // ElementDefinition.representation not present on profile.
  reserved 5;

  // Name for this particular element (in a set of slices)
  String slice_name = 6;

  // If this slice definition constrains an inherited slice definition (or not)
  Boolean slice_is_constraining = 7;

  // Name for element to display with or prompt for element
  String label = 8;

  // Corresponding codes in terminologies
  repeated Coding code = 9;

  // ElementDefinition.slicing not present on profile.
  reserved 10;

  // ElementDefinition.short not present on profile.
  reserved 11;

  // Full formal definition as narrative text
  Markdown definition = 12;

  // Comments about the use of this element
  Markdown comment = 13;

  // Why this resource has been created
  Markdown requirements = 14;

  // Other names
  repeated String alias = 15;

  // Minimum Cardinality
  UnsignedInt min = 16;

  // Maximum Cardinality (a number or *)
  String max = 17 [(.google.fhir.proto.fhir_path_constraint) =
                       "empty() or ($this = '*') or (toInteger() >= 0)"];

  // Base definition information for tools
  message Base {
    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Path that identifies the base element
    String path = 3
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Min cardinality of the base element
    UnsignedInt min = 4
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Max cardinality of the base element
    String max = 5
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];
  }
  Base base = 18;

  // ElementDefinition.contentReference not present on profile.
  reserved 19;

  // Data type and Profile for this element
  message TypeRef {
    option (.google.fhir.proto.fhir_path_message_constraint) =
        "aggregation.empty() or (code = 'Reference')";
    option (.google.fhir.proto.fhir_path_message_constraint) =
        "(code='Reference' or code = 'canonical') or targetProfile.empty()";

    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Data type or Resource (reference to definition)
    Uri code = 3
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // ElementDefinition.type.profile not present on profile.
    reserved 4;

    // Profile (StructureDefinition or IG) on the Reference/canonical target -
    // one must apply
    repeated Canonical target_profile = 5;

    // ElementDefinition.type.aggregation not present on profile.
    reserved 6;

    // either | independent | specific
    message VersioningCode {
      option (.google.fhir.proto.structure_definition_kind) =
          KIND_PRIMITIVE_TYPE;
      option (.google.fhir.proto.fhir_valueset_url) =
          "http://hl7.org/fhir/ValueSet/reference-version-rules";
      option (.google.fhir.proto.fhir_profile_base) =
          "http://hl7.org/fhir/StructureDefinition/code";

      ReferenceVersionRulesCode.Value value = 1;

      String id = 2;

      repeated Extension extension = 3;
    }
    VersioningCode versioning = 7;
  }
  repeated TypeRef type = 20;

  // Specified value if missing from instance
  message DefaultValueX {
    option (.google.fhir.proto.is_choice_type) = true;

    oneof choice {
      Base64Binary base64_binary = 1;
      Boolean boolean = 2;
      Canonical canonical = 3;
      Code code = 4;
      Date date = 5;
      DateTime date_time = 6;
      Decimal decimal = 7;
      Id id = 8;
      Instant instant = 9;
      Integer integer = 10;
      Markdown markdown = 11;
      Oid oid = 12;
      PositiveInt positive_int = 13;
      String string_value = 14 [json_name = "string"];
      Time time = 15;
      UnsignedInt unsigned_int = 16;
      Uri uri = 17;
      Url url = 18;
      Uuid uuid = 19;
      Address address = 20;
      Age age = 21;
      Annotation annotation = 22;
      Attachment attachment = 23;
      CodeableConcept codeable_concept = 24;
      Coding coding = 25;
      ContactPoint contact_point = 26;
      Count count = 27;
      Distance distance = 28;
      Duration duration = 29;
      HumanName human_name = 30;
      Identifier identifier = 31;
      Money money = 32;
      Period period = 33;
      Quantity quantity = 34;
      Range range = 35;
      Ratio ratio = 36;
      Reference reference = 37;
      SampledData sampled_data = 38;
      Signature signature = 39;
      Timing timing = 40;
      ContactDetail contact_detail = 41;
      Contributor contributor = 42;
      DataRequirement data_requirement = 43;
      Expression expression = 44;
      ParameterDefinition parameter_definition = 45;
      RelatedArtifact related_artifact = 46;
      TriggerDefinition trigger_definition = 47;
      UsageContext usage_context = 48;
      Dosage dosage = 49;
    }
  }
  DefaultValueX default_value = 21;

  // Implicit meaning when this element is missing
  Markdown meaning_when_missing = 22;

  // What the order of the elements means
  String order_meaning = 23;

  // ElementDefinition.fixed[x] not present on profile.
  reserved 24;

  // ElementDefinition.pattern[x] not present on profile.
  reserved 25;

  // Example value (as defined for type)
  message Example {
    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Describes the purpose of this example
    String label = 3
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Value of Example (one of allowed types)
    message ValueX {
      option (.google.fhir.proto.is_choice_type) = true;

      oneof choice {
        Base64Binary base64_binary = 1;
        Boolean boolean = 2;
        Canonical canonical = 3;
        Code code = 4;
        Date date = 5;
        DateTime date_time = 6;
        Decimal decimal = 7;
        Id id = 8;
        Instant instant = 9;
        Integer integer = 10;
        Markdown markdown = 11;
        Oid oid = 12;
        PositiveInt positive_int = 13;
        String string_value = 14 [json_name = "string"];
        Time time = 15;
        UnsignedInt unsigned_int = 16;
        Uri uri = 17;
        Url url = 18;
        Uuid uuid = 19;
        Address address = 20;
        Age age = 21;
        Annotation annotation = 22;
        Attachment attachment = 23;
        CodeableConcept codeable_concept = 24;
        Coding coding = 25;
        ContactPoint contact_point = 26;
        Count count = 27;
        Distance distance = 28;
        Duration duration = 29;
        HumanName human_name = 30;
        Identifier identifier = 31;
        Money money = 32;
        Period period = 33;
        Quantity quantity = 34;
        Range range = 35;
        Ratio ratio = 36;
        Reference reference = 37;
        SampledData sampled_data = 38;
        Signature signature = 39;
        Timing timing = 40;
        ContactDetail contact_detail = 41;
        Contributor contributor = 42;
        DataRequirement data_requirement = 43;
        Expression expression = 44;
        ParameterDefinition parameter_definition = 45;
        RelatedArtifact related_artifact = 46;
        TriggerDefinition trigger_definition = 47;
        UsageContext usage_context = 48;
        Dosage dosage = 49;
      }
    }
    ValueX value = 4
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];
  }
  repeated Example example = 26;

  // Minimum Allowed Value (for some types)
  message MinValueX {
    option (.google.fhir.proto.is_choice_type) = true;

    oneof choice {
      Date date = 1;
      DateTime date_time = 2;
      Instant instant = 3;
      Time time = 4;
      Decimal decimal = 5;
      Integer integer = 6;
      PositiveInt positive_int = 7;
      UnsignedInt unsigned_int = 8;
      Quantity quantity = 9;
    }
  }
  MinValueX min_value = 27;

  // Maximum Allowed Value (for some types)
  message MaxValueX {
    option (.google.fhir.proto.is_choice_type) = true;

    oneof choice {
      Date date = 1;
      DateTime date_time = 2;
      Instant instant = 3;
      Time time = 4;
      Decimal decimal = 5;
      Integer integer = 6;
      PositiveInt positive_int = 7;
      UnsignedInt unsigned_int = 8;
      Quantity quantity = 9;
    }
  }
  MaxValueX max_value = 28;

  // Max length for strings
  Integer max_length = 29;

  // Reference to invariant about presence
  repeated Id condition = 30;

  // Condition that must evaluate to true
  message Constraint {
    option (.google.fhir.proto.fhir_path_message_constraint) =
        "expression.exists()";

    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Target of 'condition' reference above
    Id key = 3 [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Why this constraint is necessary or appropriate
    String requirements = 4;

    // error | warning
    message SeverityCode {
      option (.google.fhir.proto.structure_definition_kind) =
          KIND_PRIMITIVE_TYPE;
      option (.google.fhir.proto.fhir_valueset_url) =
          "http://hl7.org/fhir/ValueSet/constraint-severity";
      option (.google.fhir.proto.fhir_profile_base) =
          "http://hl7.org/fhir/StructureDefinition/code";

      ConstraintSeverityCode.Value value = 1;

      String id = 2;

      repeated Extension extension = 3;
    }
    SeverityCode severity = 5
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Human description of constraint
    String human = 6
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // FHIRPath expression of constraint
    String expression = 7;

    // XPath expression of constraint
    String xpath = 8;

    // Reference to original source of constraint
    Canonical source = 9;
  }
  repeated Constraint constraint = 31;

  // If the element must be supported
  Boolean must_support = 32;

  // ElementDefinition.isModifier not present on profile.
  reserved 33;

  // Reason that this element is marked as a modifier
  String is_modifier_reason = 34;

  // ElementDefinition.isSummary not present on profile.
  reserved 35;

  // ValueSet details if this is coded
  message ElementDefinitionBinding {
    option (.google.fhir.proto.fhir_path_message_constraint) =
        "valueSet.exists() implies (valueSet.startsWith('http:') or valueSet.startsWith('https') or valueSet.startsWith('urn:'))";

    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // required | extensible | preferred | example
    message StrengthCode {
      option (.google.fhir.proto.structure_definition_kind) =
          KIND_PRIMITIVE_TYPE;
      option (.google.fhir.proto.fhir_valueset_url) =
          "http://hl7.org/fhir/ValueSet/binding-strength";
      option (.google.fhir.proto.fhir_profile_base) =
          "http://hl7.org/fhir/StructureDefinition/code";

      BindingStrengthCode.Value value = 1;

      String id = 2;

      repeated Extension extension = 3;
    }
    StrengthCode strength = 3
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Human explanation of the value set
    String description = 4;

    // Source of value set
    Canonical value_set = 5;
  }
  ElementDefinitionBinding binding = 36;

  // Map element to another set of definitions
  message Mapping {
    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Reference to mapping declaration
    Id identity = 3
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Computable language of mapping
    message LanguageCode {
      option (.google.fhir.proto.structure_definition_kind) =
          KIND_PRIMITIVE_TYPE;
      option (.google.fhir.proto.fhir_valueset_url) =
          "http://hl7.org/fhir/ValueSet/mimetypes";
      option (.google.fhir.proto.fhir_profile_base) =
          "http://hl7.org/fhir/StructureDefinition/code";

      // Field 1 reserved to allow enumeration in the future.
      reserved 1;

      String id = 2;

      repeated Extension extension = 3;

      // This valueset is not enumerable, and so is represented as a string.
      string value = 4;
    }
    LanguageCode language = 4;

    // Details of the mapping
    String map = 5
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Comments about the mapping or its use
    String comment = 6;
  }
  repeated Mapping mapping = 37;

  // Prompt for element phrased as question
  repeated String question = 38 [
    (.google.fhir.proto.fhir_inlined_extension_url) =
        "http://hl7.org/fhir/StructureDefinition/elementdefinition-question",
    json_name = "Question"
  ];

  // Units to use for measured value
  ElementDefinitionAllowedUnits.ValueX allowed_units = 39 [
    (.google.fhir.proto.fhir_inlined_extension_url) =
        "http://hl7.org/fhir/StructureDefinition/elementdefinition-allowedUnits",
    json_name = "AllowedUnits"
  ];
}
